crates/pathfinder/src/state.rs

```diff
@@ -24,183 +24,592 @@ use crate::{
+
+        let block_info;
+
+        match update(
+            state_update,
+            global_root,
+            &root_log,
+            &db_transaction,
+            &fetched_contracts_tx,
+            &mut ready_contracts_rx,
+        ) {
+            Ok(BlockUpdated { record, info }) => {
+                previous_state = Some(record);
+                block_info = info;
+            }
+            Err(e) => {
+                println!("erroring out");
+                return Err(anyhow::anyhow!(
+                    "Updating to block number {} gave {:?}",
+                    root_log.block_number.0,
+                    e
+                ));
+            }
+        };
+
+        db_transaction.commit().with_context(|| {
+            format!(
+                "Committing database transaction for block number {}",
+                root_log.block_number.0
             )
-            .await
-            {
-                Ok(state) => Some(state),
-                Err(UpdateError::Reorg) => todo!("Handle reorg event!"),
-                Err(UpdateError::Other(other)) => {
-                    return Err(other).with_context(|| {
-                        format!("Updating to block number {}", root_log.block_number.0)
-                    });
-                }
-            };
-            db_transaction.commit().with_context(|| {
-                format!(
-                    "Committing database transaction for block number {}",
-                    root_log.block_number.0
-                )
-            })?;
+        })?;
+
+        global_root = next_root;
 
-            global_root = root_log.global_root;
+        let progress = (
+            root_log.block_number.0,
+            std::time::Instant::now(),
+            block_info,
+        );
+
+        if block_update_tx.blocking_send(progress).is_err() {
+            break;
         }
     }
+
+    Ok(())
+}
+
+fn extract_compress(
+    tx: mpsc::Sender<FetchedCompressedContract>,
+    mut rx: mpsc::Receiver<FetchedExtractableContract>,
+) -> anyhow::Result<()> {
+    let mut compressor = zstd::bulk::Compressor::new(10)
+        .context("Couldn't create zstd compressor for ContractsTable")
+        .unwrap();
+
+    let mut process_one = |definition: bytes::Bytes| {
+        // to really reuse the buffers, we should inline the extract_abi_code_hash as well.
+        let (abi, bytecode, hash) =
+            crate::state::contract_hash::extract_abi_code_hash(&*definition)
+                .context("Compute contract hash")?;
+
+        let abi = compressor
+            .compress(&abi)
+            .context("Failed to compress ABI")?;
+        let bytecode = compressor
+            .compress(&bytecode)
+            .context("Failed to compress bytecode")?;
+        let definition = compressor
+            .compress(&*definition)
+            .context("Failed to compress definition")?;
+
+        Result::<_, anyhow::Error>::Ok(CompressedContract {
```

Comment:
Is the explicit type required here..?

Reply:
For the use of closure here it was needed at least originally.